flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+                    HsSubpartitionDataBuffer subpartitionDataBuffer =
+                            getTargetSubpartitionDataBuffer(subpartitionId);
+                    final Object subpartitionLock = subpartitionDataBuffer.getSubpartitionLock();
+                    LockUtils.runWithLock(
+                            lock.readLock(),
+                            () -> {
+                                synchronized (subpartitionLock) {
+                                    buffers.forEach(
+                                            (indexAndChannel) -> {
+                                                int bufferIndex = indexAndChannel.getBufferIndex();
+                                                BufferContext bufferContext =
+                                                        subpartitionDataBuffer.getBufferContext(
+                                                                bufferIndex);
+                                                // only spill and not consumed buffer needs to be
+                                                // marked as readable.
+                                                if (subpartitionDataBuffer.isBufferSatisfyStatus(
+                                                        bufferContext,
+                                                        SpillStatus.SPILL,
+                                                        ConsumeStatus.NOT_CONSUMED)) {
+                                                    checkNotNull(
+                                                                    bufferContext.spilledFuture,
+                                                                    "Buffer in spill status should already set spill future.")
+                                                            .thenRun(
+                                                                    () ->
+                                                                            dataIndex
+                                                                                    .markBufferReadable(
+                                                                                            subpartitionId,
+                                                                                            bufferIndex));
+                                                }
+                                                subpartitionDataBuffer.onBufferReleased(
+                                                        bufferIndex);
+                                            });
+                                }
+                            });
+                });
+    }
+
+    private void recycleBuffer(MemorySegment buffer) {
+        numRequestedBuffers.decrementAndGet();
+        bufferPool.recycle(buffer);
+    }
+
+    class HsSubpartitionDataBuffer {
+        // this field only accessed by task main thread.
+        private final Queue<BufferBuilder> unfinishedBuffers = new LinkedList<>();
```

Comment:
Why is this a queue?

Reply:
When append a record, it is sometimes necessary to request multiple `unfinished buffer`, and then fill the data on them in turn. Whenever one is filled, it will be polled out of the queue and a `finished buffer` will be created.